File: lib/cppbind/mmscenegraph/src/statistics.cpp
/*
 * Copyright (C) 2025 David Cattermole.
 *
 * This file is part of mmSolver.
 *
 * mmSolver is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * mmSolver is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with mmSolver.  If not, see <https://www.gnu.org/licenses/>.
 * ====================================================================
 *
 */

#include <mmscenegraph/_cxx.h>
#include <mmscenegraph/_cxxbridge.h>
#include <mmscenegraph/statistics.h>

namespace mmscenegraph {

MMSCENEGRAPH_API_EXPORT
Real gaussian(const Real x, const Real mean, const Real sigma) noexcept {
    return shim_gaussian(x, mean, sigma);
}

MMSCENEGRAPH_API_EXPORT
bool calc_mean_absolute_deviation(rust::Slice<const Real> &data, Real &out_mean,
                                  Real &out_mad) noexcept {
    return shim_calc_mean_absolute_deviation(data, out_mean);
}

MMSCENEGRAPH_API_EXPORT
bool calc_population_variance(rust::Slice<const Real> &data, Real &out_mean,
                              Real &out_variance) noexcept {
    return shim_calc_population_variance(data, out_mean, out_variance);
}

MMSCENEGRAPH_API_EXPORT
bool calc_sample_variance(rust::Slice<const Real> &data, Real &out_mean,
                          Real &out_variance) noexcept {
    return shim_calc_sample_variance(data, out_mean, out_variance);
}

MMSCENEGRAPH_API_EXPORT
bool calc_population_standard_deviation(rust::Slice<const Real> &data,
                                        Real &out_mean,
                                        Real &out_std_dev) noexcept {
    return shim_calc_population_standard_deviation(data, out_mean, out_std_dev);
}

MMSCENEGRAPH_API_EXPORT
bool calc_sample_standard_deviation(rust::Slice<const Real> &data,
                                    Real &out_mean,
                                    Real &out_std_dev) noexcept {
    return shim_calc_sample_standard_deviation(data, out_mean, out_std_dev);
}

MMSCENEGRAPH_API_EXPORT
bool calc_population_coefficient_of_variation(rust::Slice<const Real> &data,
                                              Real &out_mean,
                                              Real &out_cv) noexcept {
    return shim_calc_population_coefficient_of_variation(data, out_mean,
                                                         out_cv);
}

MMSCENEGRAPH_API_EXPORT
bool calc_sample_coefficient_of_variation(rust::Slice<const Real> &data,
                                          Real &out_mean,
                                          Real &out_cv) noexcept {
    return shim_calc_sample_coefficient_of_variation(data, out_mean, out_cv);
}

MMSCENEGRAPH_API_EXPORT
bool calc_population_relative_standard_deviation(rust::Slice<const Real> &data,
                                                 Real &out_mean,
                                                 Real &out_rsd) noexcept {
    return shim_calc_population_relative_standard_deviation(data, out_mean,
                                                            out_rsd);
}

MMSCENEGRAPH_API_EXPORT
bool calc_sample_relative_standard_deviation(rust::Slice<const Real> &data,
                                             Real &out_mean,
                                             Real &out_rsd) noexcept {
    return shim_calc_sample_relative_standard_deviation(data, out_mean,
                                                        out_rsd);
}

MMSCENEGRAPH_API_EXPORT
Real calc_z_score(const Real mean, const Real std_dev,
                  const Real value) noexcept {
    return shim_calc_z_score(mean, std_dev, value);
}

MMSCENEGRAPH_API_EXPORT
bool calc_peak_to_peak(rust::Slice<const Real> &data,
                       Real &out_value) noexcept {
    return shim_calc_peak_to_peak(data, out_value);
}

MMSCENEGRAPH_API_EXPORT
bool calc_skewness_type1(rust::Slice<const Real> &data, Real &out_mean,
                         Real &out_skewness) noexcept {
    return shim_calc_skewness_type1(data, out_mean, out_skewness);
}

MMSCENEGRAPH_API_EXPORT
bool calc_skewness_type2(rust::Slice<const Real> &data, Real &out_mean,
                         Real &out_skewness) noexcept {
    return shim_calc_skewness_type2(data, out_mean, out_skewness);
}

MMSCENEGRAPH_API_EXPORT
bool calc_population_kurtosis_excess(rust::Slice<const Real> &data,
                                     Real &out_mean,
                                     Real &out_kurtosis) noexcept {
    return shim_calc_population_kurtosis_excess(data, out_mean, out_kurtosis);
}

MMSCENEGRAPH_API_EXPORT
bool calc_sample_kurtosis_excess(rust::Slice<const Real> &data, Real &out_mean,
                                 Real &out_kurtosis) noexcept {
    return shim_calc_sample_kurtosis_excess(data, out_mean, out_kurtosis);
}

MMSCENEGRAPH_API_EXPORT
bool calc_local_minima_maxima(rust::Slice<const Real> &data,
                              rust::Vec<size_t> &out_indices) noexcept {
    return shim_calc_local_minima_maxima(data, out_indices);
}

MMSCENEGRAPH_API_EXPORT
bool calc_signal_to_noise_ratio(rust::Slice<const Real> &data, Real &out_mean,
                                Real &out_snr) noexcept {
    return shim_calc_signal_to_noise_ratio(data, out_mean, out_snr);
}

MMSCENEGRAPH_API_EXPORT
bool calc_signal_to_noise_ratio_as_decibels(rust::Slice<const Real> &data,
                                            Real &out_mean,
                                            Real &out_snr_db) noexcept {
    return shim_calc_signal_to_noise_ratio_as_decibels(data, out_mean,
                                                       out_snr_db);
}

MMSCENEGRAPH_API_EXPORT
bool calc_median_absolute_deviation(rust::Slice<const Real> &sorted_data,
                                    Real &out_median, Real &out_mad) noexcept {
    return shim_calc_median_absolute_deviation(sorted_data, out_median,
                                               out_mad);
}

MMSCENEGRAPH_API_EXPORT
bool calc_median_absolute_deviation_sigma(const Real value,
                                          rust::Slice<const Real> &sorted_data,
                                          Real &out_median,
                                          Real &out_sigma) noexcept {
    return shim_calc_median_absolute_deviation_sigma(value, sorted_data,
                                                     out_median, out_sigma);
}

MMSCENEGRAPH_API_EXPORT
bool calc_quantile(rust::Slice<const Real> &sorted_data, const Real probability,
                   Real &out_value) noexcept {
    return shim_calc_quantile(sorted_data, probability, out_value);
}

MMSCENEGRAPH_API_EXPORT
bool calc_quartiles(rust::Slice<const Real> &sorted_data, Real &out_q1,
                    Real &out_q2, Real &out_q3) noexcept {
    return shim_calc_quartiles(sorted_data, out_q1, out_q2, out_q3);
}

MMSCENEGRAPH_API_EXPORT
bool calc_interquartile_range(rust::Slice<const Real> &sorted_data,
                              Real &out_median, Real &out_iqr) noexcept {
    return shim_calc_interquartile_range(sorted_data, out_median, out_iqr);
}

MMSCENEGRAPH_API_EXPORT
bool calc_percentile_rank(rust::Slice<const Real> &sorted_data,
                          const Real value, Real &out_rank) noexcept {
    return shim_calc_percentile_rank(sorted_data, value, out_rank);
}

}  // namespace mmscenegraph

File: lib/cppbind/mmscenegraph/src/statistics.rs
//
// Copyright (C) 2025 David Cattermole.
//
// This file is part of mmSolver.
//
// mmSolver is free software: you can redistribute it and/or modify it
// under the terms of the GNU Lesser General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// mmSolver is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with mmSolver.  If not, see <https://www.gnu.org/licenses/>.
// ====================================================================
//

use mmscenegraph_rust::constant::Real as CoreReal;
use mmscenegraph_rust::math::statistics::{
    calc_coefficient_of_determination as core_calc_coefficient_of_determination,
    calc_interquartile_range as core_calc_interquartile_range,
    calc_local_minima_maxima as core_calc_local_minima_maxima,
    calc_mean_absolute_deviation as core_calc_mean_absolute_deviation,
    calc_mean_absolute_error as core_calc_mean_absolute_error,
    calc_median_absolute_deviation as core_calc_median_absolute_deviation,
    calc_median_absolute_deviation_sigma as core_calc_median_absolute_deviation_sigma,
    calc_normalized_root_mean_square_error as core_calc_normalized_root_mean_square_error,
    calc_peak_to_peak as core_calc_peak_to_peak,
    calc_percentile_rank as core_calc_percentile_rank,
    calc_population_coefficient_of_variation as core_calc_population_coefficient_of_variation,
    calc_population_kurtosis_excess as core_calc_population_kurtosis_excess,
    calc_population_relative_standard_deviation as core_calc_population_relative_standard_deviation,
    calc_population_standard_deviation as core_calc_population_standard_deviation,
    calc_population_variance as core_calc_population_variance,
    calc_quantile as core_calc_quantile, calc_quartiles as core_calc_quartiles,
    calc_root_mean_square_error as core_calc_root_mean_square_error,
    calc_sample_coefficient_of_variation as core_calc_sample_coefficient_of_variation,
    calc_sample_kurtosis_excess as core_calc_sample_kurtosis_excess,
    calc_sample_relative_standard_deviation as core_calc_sample_relative_standard_deviation,
    calc_sample_standard_deviation as core_calc_sample_standard_deviation,
    calc_sample_variance as core_calc_sample_variance,
    calc_signal_to_noise_ratio as core_calc_signal_to_noise_ratio,
    calc_signal_to_noise_ratio_as_decibels as core_calc_signal_to_noise_ratio_as_decibels,
    calc_skewness_type1 as core_calc_skewness_type1,
    calc_skewness_type2 as core_calc_skewness_type2,
    calc_z_score as core_calc_z_score, gaussian as core_gaussian,
    SortedDataSlice, SortedDataSliceOps, UnsortedDataSlice,
    UnsortedDataSliceOps,
};

pub fn shim_gaussian(x: CoreReal, mean: CoreReal, sigma: CoreReal) -> CoreReal {
    core_gaussian(x, mean, sigma)
}

pub fn shim_calc_mean_absolute_deviation(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_mean_absolute_deviation(&data_slice) {
                Ok(_mad) => {
                    *out_mean = data_slice.mean();
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_population_variance(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_variance: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => match core_calc_population_variance(&data_slice) {
            Ok(variance) => {
                *out_mean = data_slice.mean();
                *out_variance = variance;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_sample_variance(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_variance: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => match core_calc_sample_variance(&data_slice) {
            Ok(variance) => {
                *out_mean = data_slice.mean();
                *out_variance = variance;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_population_standard_deviation(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_std_dev: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_population_standard_deviation(&data_slice) {
                Ok(std_dev) => {
                    *out_mean = data_slice.mean();
                    *out_std_dev = std_dev;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_sample_standard_deviation(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_std_dev: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_sample_standard_deviation(&data_slice) {
                Ok(std_dev) => {
                    *out_mean = data_slice.mean();
                    *out_std_dev = std_dev;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_population_coefficient_of_variation(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_cv: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_population_coefficient_of_variation(&data_slice) {
                Ok(cv) => {
                    *out_mean = data_slice.mean();
                    *out_cv = cv;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_sample_coefficient_of_variation(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_cv: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_sample_coefficient_of_variation(&data_slice) {
                Ok(cv) => {
                    *out_mean = data_slice.mean();
                    *out_cv = cv;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_population_relative_standard_deviation(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_rsd: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_population_relative_standard_deviation(&data_slice)
            {
                Ok(rsd) => {
                    *out_mean = data_slice.mean();
                    *out_rsd = rsd;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_sample_relative_standard_deviation(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_rsd: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_sample_relative_standard_deviation(&data_slice) {
                Ok(rsd) => {
                    *out_mean = data_slice.mean();
                    *out_rsd = rsd;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_z_score(
    mean: CoreReal,
    std_dev: CoreReal,
    value: CoreReal,
) -> CoreReal {
    core_calc_z_score(mean, std_dev, value)
}

pub fn shim_calc_peak_to_peak(
    data: &[CoreReal],
    out_value: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => match core_calc_peak_to_peak(&data_slice) {
            Ok(value) => {
                *out_value = value;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_skewness_type1(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_skewness: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => match core_calc_skewness_type1(&data_slice) {
            Ok(skewness) => {
                *out_mean = data_slice.mean();
                *out_skewness = skewness;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_skewness_type2(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_skewness: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => match core_calc_skewness_type2(&data_slice) {
            Ok(skewness) => {
                *out_mean = data_slice.mean();
                *out_skewness = skewness;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_population_kurtosis_excess(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_kurtosis: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_population_kurtosis_excess(&data_slice, None) {
                Ok(kurtosis) => {
                    *out_mean = data_slice.mean();
                    *out_kurtosis = kurtosis;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_sample_kurtosis_excess(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_kurtosis: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_sample_kurtosis_excess(&data_slice, None) {
                Ok(kurtosis) => {
                    *out_mean = data_slice.mean();
                    *out_kurtosis = kurtosis;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_local_minima_maxima(
    data: &[CoreReal],
    out_indices: &mut Vec<usize>,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => match core_calc_local_minima_maxima(&data_slice) {
            Ok(indices) => {
                *out_indices = indices;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_signal_to_noise_ratio(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_snr: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => match core_calc_signal_to_noise_ratio(&data_slice) {
            Ok(snr) => {
                *out_mean = data_slice.mean();
                *out_snr = snr;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_signal_to_noise_ratio_as_decibels(
    data: &[CoreReal],
    out_mean: &mut CoreReal,
    out_snr_db: &mut CoreReal,
) -> bool {
    match UnsortedDataSlice::new(data, None) {
        Ok(data_slice) => {
            match core_calc_signal_to_noise_ratio_as_decibels(&data_slice) {
                Ok(snr_db) => {
                    *out_mean = data_slice.mean();
                    *out_snr_db = snr_db;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_median_absolute_deviation(
    sorted_data: &[CoreReal],
    out_median: &mut CoreReal,
    out_mad: &mut CoreReal,
) -> bool {
    match SortedDataSlice::new(sorted_data, None, None) {
        Ok(data_slice) => {
            let mut sorted_deviations = vec![0.0; sorted_data.len()];
            match core_calc_median_absolute_deviation(
                &data_slice,
                &mut sorted_deviations,
            ) {
                Ok(mad) => {
                    *out_median = data_slice.median();
                    *out_mad = mad;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_median_absolute_deviation_sigma(
    value: CoreReal,
    sorted_data: &[CoreReal],
    out_median: &mut CoreReal,
    out_sigma: &mut CoreReal,
) -> bool {
    match SortedDataSlice::new(sorted_data, None, None) {
        Ok(data_slice) => {
            let mut sorted_deviations = vec![0.0; sorted_data.len()];
            match core_calc_median_absolute_deviation_sigma(
                value,
                &data_slice,
                &mut sorted_deviations,
            ) {
                Ok(sigma) => {
                    *out_median = data_slice.median();
                    *out_sigma = sigma;
                    true
                }
                Err(_) => false,
            }
        }
        Err(_) => false,
    }
}

pub fn shim_calc_quantile(
    sorted_data: &[CoreReal],
    probability: CoreReal,
    out_value: &mut CoreReal,
) -> bool {
    match SortedDataSlice::new(sorted_data, None, None) {
        Ok(data_slice) => match core_calc_quantile(&data_slice, probability) {
            Ok(value) => {
                *out_value = value;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_quartiles(
    sorted_data: &[CoreReal],
    out_q1: &mut CoreReal,
    out_q2: &mut CoreReal,
    out_q3: &mut CoreReal,
) -> bool {
    match SortedDataSlice::new(sorted_data, None, None) {
        Ok(data_slice) => match core_calc_quartiles(&data_slice) {
            Ok((q1, q2, q3)) => {
                *out_q1 = q1;
                *out_q2 = q2;
                *out_q3 = q3;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_interquartile_range(
    sorted_data: &[CoreReal],
    out_median: &mut CoreReal,
    out_iqr: &mut CoreReal,
) -> bool {
    match SortedDataSlice::new(sorted_data, None, None) {
        Ok(data_slice) => match core_calc_interquartile_range(&data_slice) {
            Ok(iqr) => {
                *out_median = data_slice.median();
                *out_iqr = iqr;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_percentile_rank(
    sorted_data: &[CoreReal],
    value: CoreReal,
    out_rank: &mut CoreReal,
) -> bool {
    match SortedDataSlice::new(sorted_data, None, None) {
        Ok(data_slice) => match core_calc_percentile_rank(&data_slice, value) {
            Ok(rank) => {
                *out_rank = rank;
                true
            }
            Err(_) => false,
        },
        Err(_) => false,
    }
}

pub fn shim_calc_mean_absolute_error(
    actual: &[CoreReal],
    predicted: &[CoreReal],
    out_mae: &mut CoreReal,
) -> bool {
    match core_calc_mean_absolute_error(actual, predicted) {
        Ok(mae) => {
            *out_mae = mae;
            true
        }
        Err(_) => false,
    }
}

pub fn shim_calc_root_mean_square_error(
    actual: &[CoreReal],
    predicted: &[CoreReal],
    out_rmse: &mut CoreReal,
) -> bool {
    match core_calc_root_mean_square_error(actual, predicted) {
        Ok(rmse) => {
            *out_rmse = rmse;
            true
        }
        Err(_) => false,
    }
}

pub fn shim_calc_normalized_root_mean_square_error(
    actual: &[CoreReal],
    predicted: &[CoreReal],
    out_nrmse: &mut CoreReal,
) -> bool {
    match core_calc_normalized_root_mean_square_error(actual, predicted) {
        Ok(nrmse) => {
            *out_nrmse = nrmse;
            true
        }
        Err(_) => false,
    }
}

pub fn shim_calc_coefficient_of_determination(
    actual: &[CoreReal],
    predicted: &[CoreReal],
    out_r_squared: &mut CoreReal,
) -> bool {
    match core_calc_coefficient_of_determination(actual, predicted) {
        Ok(r_squared) => {
            *out_r_squared = r_squared;
            true
        }
        Err(_) => false,
    }
}
